### ISSUE TITLE
Pinecil v2 doesn't work after updating to v2.21

### ISSUE THREAD

**Issue by sorgelig** [Ralim/IronOS#1661]
**Describe the bug**
After flashing by blisp to v2.21 my pinecil v2 looks completely dead.

**To Reproduce**
1. Flash v2.21 (Pinecilv2_EN.bin)
2. iron looks dead, no display on screen

**Expected behavior**
normal work

**Details of your device:**
 
 - Device: Pinecil v2 bought on pine64 site, so should be original one.
 - Release: i don't know how to see. in flash mode it doesn't show anything on screen.
 - Power adapter being used: tried both Type-C and 5525

**Additional context**
v2.21 it seems doesn't work on my Pinecil v2. Thankfully, flashing v2.20 brings iron back to life.

**Comment by sorgelig** [comment]
**Update:** just passed authenticity test (just for sanity). Manufacture date: February 27, 2023.
Not sure if it's ok to post ID and SN here, but should be irrelevant to issue as long as it's original.

**Comment by gamelaster** [comment]
No need to post it here. This looks like software issue, as rollback to v2.20 helps.

**Comment by sorgelig** [comment]
There is no sign of exactly software/updater issue. Broken firmware also can lead to such state.

**Comment by River-Mochi** [comment]
> There is no sign of exactly software/updater issue. Broken firmware also can lead to such state. Ability to flash any version of firmware comes from bootloader, and has nothing to do with software.

What  OS  are you using for the update  and type of computer?

**Comment by River-Mochi** [comment]
Is  there any possibility   you might have accidentally not used the BIN  file when you updated?

if for example  you   flash the 2.21  **DFU**  or HEX   file , then you will get black screen also.   I did this before and I had to flash again with BIN  file to fix it. 

try it one more time  to be sure and make sure to follow instructions in Wiki page.   You can't really brick the pinecil doing this so it is okay to try  2-3  times.  

https://wiki.pine64.org/wiki/Pinecil_Firmware#Update_V2:_Windows

**Comment by gamelaster** [comment]
@sorgelig by software I mean firmware, but I think it can be also flasher issue.
Older `blisp` version had issue with bigger firmwares (those with Bluetooth), so be sure you have latest blisp.

**Comment by River-Mochi** [comment]
Get the **newest** BLISP here.  **https://github.com/pine64/blisp/releases/tag/v0.0.3**

updated  two Pinecil V2, no problems with 2.21  so far.   other people also  in pinecil chat say it's good for them. 
<br>
<img src="https://user-images.githubusercontent.com/97197236/230780659-1837814a-80e0-45fe-b69b-94cfb1635716.png" width="400">

**Comment by River-Mochi** [comment]
> Older `blisp` version had issue with bigger firmwares (those with Bluetooth), so be sure you have latest blisp.

Thanks for this information, we already have bunch of people updating this morning with different issues. so far all resolved and Success in the end.

**Comment by sorgelig** [comment]
I'm using blisp 0.0.3. Windows 11.
Yes, i've tried Pinecilv2_EN.bin and also tried multi-language version bin - both give non-working iron. Also i've tried dev version which i've downloaded several days ago - also black screen. 
Only flashing v2.20 brings iron to life.

I don't know what else i'm missing...

**Comment by River-Mochi** [comment]
do you  mean you also  tried the earlier  Beta a few days ago and also got a black screen?

**Comment by sorgelig** [comment]
```
blisp.exe write -c bl70x --reset .\Pinecilv2_EN.bin
Sending a handshake...
Handshake successful!
Getting chip info...
BootROM version 1.0.2.7, ChipID: 00003DF5221017A8
0b / 59200b (0.00%)
4092b / 59200b (6.91%)
8184b / 59200b (13.82%)
12276b / 59200b (20.74%)
16368b / 59200b (27.65%)
20460b / 59200b (34.56%)
24552b / 59200b (41.47%)
28644b / 59200b (48.39%)
32736b / 59200b (55.30%)
36828b / 59200b (62.21%)
40920b / 59200b (69.12%)
45012b / 59200b (76.03%)
49104b / 59200b (82.95%)
53196b / 59200b (89.86%)
57288b / 59200b (96.77%)
59200b / 59200b (100.00%)
Sending a handshake...
Handshake with eflash_loader successful.
Erasing flash, this might take a while...
Flashing boot header...
Flashing the firmware...
0b / 189960b (0.00%)
2052b / 189960b (1.08%)
4104b / 189960b (2.16%)
6156b / 189960b (3.24%)
8208b / 189960b (4.32%)
10260b / 189960b (5.40%)
12312b / 189960b (6.48%)
14364b / 189960b (7.56%)
16416b / 189960b (8.64%)
18468b / 189960b (9.72%)
20520b / 189960b (10.80%)
22572b / 189960b (11.88%)
24624b / 189960b (12.96%)
26676b / 189960b (14.04%)
28728b / 189960b (15.12%)
30780b / 189960b (16.20%)
32832b / 189960b (17.28%)
34884b / 189960b (18.36%)
36936b / 189960b (19.44%)
38988b / 189960b (20.52%)
41040b / 189960b (21.60%)
43092b / 189960b (22.68%)
45144b / 189960b (23.77%)
47196b / 189960b (24.85%)
49248b / 189960b (25.93%)
51300b / 189960b (27.01%)
53352b / 189960b (28.09%)
55404b / 189960b (29.17%)
57456b / 189960b (30.25%)
59508b / 189960b (31.33%)
61560b / 189960b (32.41%)
63612b / 189960b (33.49%)
65664b / 189960b (34.57%)
67716b / 189960b (35.65%)
69768b / 189960b (36.73%)
71820b / 189960b (37.81%)
73872b / 189960b (38.89%)
75924b / 189960b (39.97%)
77976b / 189960b (41.05%)
80028b / 189960b (42.13%)
82080b / 189960b (43.21%)
84132b / 189960b (44.29%)
86184b / 189960b (45.37%)
88236b / 189960b (46.45%)
90288b / 189960b (47.53%)
92340b / 189960b (48.61%)
94392b / 189960b (49.69%)
96444b / 189960b (50.77%)
98496b / 189960b (51.85%)
100548b / 189960b (52.93%)
102600b / 189960b (54.01%)
104652b / 189960b (55.09%)
106704b / 189960b (56.17%)
108756b / 189960b (57.25%)
110808b / 189960b (58.33%)
112860b / 189960b (59.41%)
114912b / 189960b (60.49%)
116964b / 189960b (61.57%)
119016b / 189960b (62.65%)
121068b / 189960b (63.73%)
123120b / 189960b (64.81%)
125172b / 189960b (65.89%)
127224b / 189960b (66.97%)
129276b / 189960b (68.05%)
131328b / 189960b (69.13%)
133380b / 189960b (70.21%)
135432b / 189960b (71.30%)
137484b / 189960b (72.38%)
139536b / 189960b (73.46%)
141588b / 189960b (74.54%)
143640b / 189960b (75.62%)
145692b / 189960b (76.70%)
147744b / 189960b (77.78%)
149796b / 189960b (78.86%)
151848b / 189960b (79.94%)
153900b / 189960b (81.02%)
155952b / 189960b (82.10%)
158004b / 189960b (83.18%)
160056b / 189960b (84.26%)
162108b / 189960b (85.34%)
164160b / 189960b (86.42%)
166212b / 189960b (87.50%)
168264b / 189960b (88.58%)
170316b / 189960b (89.66%)
172368b / 189960b (90.74%)
174420b / 189960b (91.82%)
176472b / 189960b (92.90%)
178524b / 189960b (93.98%)
180576b / 189960b (95.06%)
182628b / 189960b (96.14%)
184680b / 189960b (97.22%)
186732b / 189960b (98.30%)
188784b / 189960b (99.38%)
189960b / 189960b (100.00%)
Checking program...
Program OK!
Resetting the chip.
Flash complete!
```

**Comment by sorgelig** [comment]
> do you mean you also tried the earlier Beta a few days ago and also got a black screen?

yes

**Comment by River-Mochi** [comment]
Could you  try this  old   Beta BIN   with Bluetooth from   Nov  2022 that  Joric still has on his website.

https://github.com/joric/pinecil/wiki

**Comment by sorgelig** [comment]
> Could you try this old Beta BIN with Bluetooth from Nov 2022 that Joric still has on his website.
> This is pre-2.20 stable but has bluetooth.

doesn't work too.

**Comment by River-Mochi** [comment]
Now we know you tried  Pre-2.20  with bluetooth(nov 2022 beta),  some of the 2.20xxxx  with bluetooth and  2.21 with bluetooth all cause  black screen.  And only  2.20 with No bluetooth on it works for you. 

That is the extent of my  knoweledge of what to try. need to wait for Ralim and Gamiee on this.

**Comment by River-Mochi** [comment]
One more thing to Try...
I'm insterested if you install   2.21  black screen is fine.   then open PineSAM  on browser,  do you see information and  Debug stuff?  

2.21  all all those beta's should have Bluetooth on by default 

Pinesam here:   https://github.com/builder555/PineSAM   Download  the Premade =binary for windows and follow instructions under Premade Binaries.

**Comment by River-Mochi** [comment]
watch the terminal windows when you  connect  to Pinesam. has important messages like if it Finds your pinecil  and that you can post screen shot .

**Comment by sorgelig** [comment]
btw, about  Joric's pre-2.20 - it's not really pre.. it's pointed to releases/latest which currently is 2.21.

**Comment by sorgelig** [comment]
I discovered, on v2.21, if i press button+ while connecting USB then sometimes iron starts (with debug screen) and then it works normally. It only happens sometimes, not every time. So i have to try several times to make it work. Without button, it never starts.
So, for me it looks like some booting process is buggy, it doesn't wait enough for HW to initialize and then get stuck.

**Comment by sorgelig** [comment]
> Other thing is if you can get Pinesam up, Reset all Settings to Defaults. do that in 2.20 also.

i've already tried resetting before. It doesn't help.
Not sure if i want to mess with pineSAM. It involves additional variables, i will need to find a working BT dongle, etc... I don't see how it may help.

**Comment by sorgelig** [comment]
Partial success! Once i've got it working through button+ i've disabled BT. Now it starts without problems and without button.
Basically i don't need bluetooth but i think this bug should be fixed. I don't know if my iron has broken BT or startup time is too long to make booting process hang.

**Comment by sorgelig** [comment]
Some more info (may be useful): if i enable BT and then after several tries able to boot into debug mode, then my phone sees Pinecil BT device, so BLE is working on my iron. Just startup process somewhere hangs (when it doesn't start).

**Comment by River-Mochi** [comment]
> > Other thing is if you can get Pinesam up, Reset all Settings to Defaults. do that in 2.20 also.
> 
> i've already tried resetting before. It doesn't help. Not sure if i want to mess with pineSAM. It involves additional variables, i will need to find a working BT dongle, etc... I don't see how it may help.

Actually Pinesam doesn't  you can be up and running in 1 min
you just install that Binary which opens a browser window and you are done.  it runs from browser.

**Comment by River-Mochi** [comment]
The difference with PineSam  is it shows  you all of the Information all the Debug, allows all setting changes too.  Even if your  screen  is buggy, it will  let you control Pinecil.  
if you press (+)  to heat up,   then   you can see on pineSAM it heats up. 

the premade binaries,  it's  1  min  to execute it and get it going with   ./start script.   on some windows PC with extra security  you have to do command in Powershell to allow  Powershell to run scripts.   

Powershell runs on MacOS, Linux and Windows but Only windows  locks it down from doing script until  you personally set the permission  which is odd since if I'm opening powershell, I want to run a script.

**Comment by River-Mochi** [comment]
> btw, about Joric's pre-2.20 - it's not really pre.. it's pointed to releases/latest which currently is 2.21.

you are right he updated it recently  to  29-03-23 version.

**Comment by sorgelig** [comment]
PineSAM won't help. Please read my posts above. Iron really stuck in booting somewhere in BT initialization, so BT doesn't work, buttons don't work (no heat-up on button press). Disabling BT make my iron boot and work normally.

**Comment by River-Mochi** [comment]
> 

oic  you don't have any  Bluetooth in your PC or laptop already then .    nice thing about Pinesam is  simply  if you have a black screen it tells us if the rest of the Pinecil is still working.

we have one person who messed up his  screen and is still using Pinecil completely from PineSAM on his phone.

**Comment by River-Mochi** [comment]
> PineSAM won't help. Please read my posts above. Iron really stuck in booting somewhere in BT initialization, so BT doesn't work, buttons don't work (no heat-up on button press). Disabling BT make my iron boot and work normally.

Since you are able to partially get  2.21 started by pressing the (+)  button  go to Advanced menu, and turn off BLuetooth. and then see if it boots normally.

**Comment by sorgelig** [comment]
> Since you are able to partially get 2.21 started by pressing the (+) button go to Advanced menu, and turn off BLuetooth. and then see if it boots normally.

please read my posts above. It's already done and working.

**Comment by River-Mochi** [comment]
do you think it's the firmware or hardware? if you are inside warranty period perhaps they will exchange it.

since it's working for people on bluetooth even for the betas over past several months.

**Comment by sorgelig** [comment]
I think it's a bug in firmware. I'm also HW and SW developer, and for me it looks like initialization process doesn't wait for some states.

**Comment by River-Mochi** [comment]
I'm on a small screen right now outside with doggies.  

Hopefully Ralim has work around for this .  if it's defective they should honor a replacement  imho since this is first stable firmware that exposes BLE.

and first case  I've seen in with beta bLE over months of people using the  Beta BLE  where screen is black or partially unstable when ble enabled.   could  poor soldering  cold joint  of some part internally caused this unstable behavior. 

would be interesting to see over next few weeks if other people report on this or this is unique.

**Comment by chiefos** [comment]
Yeah, have the same issue, reconnecting the cable sometimes helps to start Pinecil , sometimes not. My opinion  - it isn`t caused by BLE, maybe PPS/PD negotiating issue? Hmm..

**Comment by PSandro** [comment]
I can confirm this for my pinecil v2. I just upgraded to the latest firmware. Sometimes when I plug in the USB-C cable, a black screen is shown and nothing happens. After replugging the cable, everything works fine usually. Sometimes it needs another replug. 

It also works in the opposite way: I plug in the usb-c cable - everything works fine; I replug the cable and the bug appears again.

**Comment by River-Mochi** [comment]
> My opinion - it isn`t caused by BLE, maybe PPS/PD negotiating issue? Hmm..

there was a PPS  change in 2.21.    have you  used  any of the other bluetooth betas before this one?

**Comment by River-Mochi** [comment]
I just had a weird issue. 
 Pinepower Portible charger PD20V/65w (newer model from ~Sept 2022).  I plug in cable and screen is black.  I unplug  and plug in again and it's fine the full 20V.  I don't know if it's the charger or the firmware.   
 The charger was plugged in the power strip for a couple hours and unused when I  tried it, don't know if this gan II charger goes into a low energy mode so that it doesn't drain a lot of electricity while sitting unused.
- however,  trying to unplug and plug in again 10 times  I can not repeat the black screen.

**Comment by chiefos** [comment]
> 

I tried some betas with ble before. Had the same issue with them. 
I`m using Baseus Blade Laptop Power Bank 100W and Baseus 100W GaN USB Charger Desktop Power Strip with Baseus E-marker chip cables and also with Pine cables from the pine64.com

**Comment by River-Mochi** [comment]
Tested my  2nd  V2,  and got a Reboot  on plugging it in and now it's normal. that is unusual, the reboot.  Same 20V PD  Pinepower portable charger.

**Comment by River-Mochi** [comment]
also  notice  that when it does work, I can see it  going from 5V increments to 20V  on detailed screen.  Normally I'm used to it being a lot faster jumping to 20V -  so fast I can't normally see all the little incremental steps on the screen from 5v.1v  and all the little decimals up to  19. 9V.

**Comment by sorgelig** [comment]
Nothing to do with PPS. I test with USB from PC. So it's pure 5V-only.

**Comment by River-Mochi** [comment]
Another test unplugging and plugging into   Pinepower desktop 120W several times,  it reboots a couple of the times  (not solid black like my other test but did reboot  with enough repeats). 
keep volume up to know which time i'm just unplugging versus which time  I'm  getting a reboot. 

https://user-images.githubusercontent.com/97197236/230795420-44fbf637-b480-4bd3-8352-09b14eb86155.mov

**Comment by River-Mochi** [comment]
I turned **off**  BLuetooth, and  did same test, and I _still get the Reboots_ after so many unplug+ plug in.  using same charger Pinepower desktop 120w (to be consisitant).

**Comment by sorgelig** [comment]
It doesn't look like reboot. It's power cut while plugging. You can see even your power meter turns off. It can be just behavior of your PSU's MCU. MCU in PSU handles power distribution among sockets. When you plug/unplug quickly, the MCU in PSU may get some spikes and unfinished PD/QC negotiation which may lead MCU to reset the circuit and cut the power temporarily.

**Comment by River-Mochi** [comment]
> It doesn't look like reboot. It's power cut while plugging. You can see even your power meter turns off. It can be just behavior of your PSU's MCU. MCU in PSU handles power distribution among sockets. When you plug/unplug quickly, the MCU in PSU may get some spikes and unfinished PD/QC negotiation which may lead MCU to reset the circuit and cut the power temporarily.

 even if I just plug in first time sometimes and leave it pluged it  it reboots.  I only did the Video  a bunch of times to show it more.  I could just do a video showing it one time also. 
Example.  leaving Pinecil unpluged for  10 min.  then plugging it in first time can also reboot.

**Comment by PSandro** [comment]
> > It doesn't look like reboot. It's power cut while plugging. You can see even your power meter turns off. It can be just behavior of your PSU's MCU. MCU in PSU handles power distribution among sockets. When you plug/unplug quickly, the MCU in PSU may get some spikes and unfinished PD/QC negotiation which may lead MCU to reset the circuit and cut the power temporarily.
> 
> even if I just plug in first time sometimes and leave it pluged it it reboots. I only did the Video a bunch of times to show it more. I could just do a video showing it one time also. Example. leaving Pinecil unpluged for 10 min. then plugging it in first time can also reboot.

that sounds like a different issue to me. I'm only experiencing the blackscreen

**Comment by River-Mochi** [comment]
> that sounds like a different issue to me. I'm only experiencing the blackscreen

I also get the **solid  black screen** , see my photo above , pressing buttons does not do anything.   this doesn't happen as often on testing. and   Rebooting  seems to fix it.    That photo I took after just leaving it like that for over 3 min. to see if anything would change.  but it just stayed like that.

**Comment by River-Mochi** [comment]
I'll try to find my 3rd V2 downstairs and see if there is a difference.

**Comment by PSandro** [comment]
> @PSandro
> 
> 1. have you tried testing with bluetooth off?
> 2. have you tried   any of the earlier  beta with bluetooth, do they also cause black screen?
I can confirm that turning bluetooth of in the settings fixes the issue. After that I didn't get any blackscreens and the pinecilv2 functioned normally. 

I haven;t tried any earlier beta versions. Before v2.21 I was running v2.20. I didn't experience any issues there.

**Comment by Ralim** [comment]
Hmm, 
By any chance do either of you have the breakout board?
Being able to see what the uart says would be super helpful here. I'm concerned it's a chip hard crash, which can be messy to debug without the uart output. As I have basically zero support on the BLE stack I would not be overly surprised if it behaves badly with "something".

I have not yet been able to reproduce this on any of my power sources or my test irons.

I can try making a build without Bluetooth to try and narrow it down ? Open to other ideas of how to debug without asking you to flash and test all the builds during 2.21.

**Comment by sorgelig** [comment]
@Ralim 
I don't think i want to open iron, i have bad experience in opening TS100... But i'm ready to try as many builds as you want. 
If you have a zip file with set of builds, then i'm ready to go through that.

**Comment by PSandro** [comment]
> Hmm,\nBy any chance do either of you have the breakout board?

I was also thinking about the UART. I ordered the Breakout Board, yesterday but I could take some time to arrive.

**Comment by Ralim** [comment]
No requirement to open the iron, just access to some form of USB C breakout. To get the uart lines.
I'll try and make some firmware to test when I can

**Comment by sorgelig** [comment]
> No requirement to open the iron, just access to some form of USB C breakout. To get the uart lines. I'll try and make some firmware to test when I can

Please provide link to pinout and connection parameters. I think i can get breakout board today or tomorrow.

**Comment by PSandro** [comment]
> Please provide link to pinout and connection parameters. I think i can get breakout board today or tomorrow.

There's information for the Pinecil Breakout board in the [pine64 wiki](https://wiki.pine64.org/wiki/Pinecil#Pinecil_Breakout_board) including a [board schematic](https://files.pine64.org/doc/Pinecil/Pinecil_Breakout_Board_Schematic_v1.0_20201005.pdf) and [PDB layout](https://files.pine64.org/doc/Pinecil/Pinecil_Breakout_Board_PCB_layout_v1.0_20201005.pdf).

There's also an image on how to connect the breakout board to an UART ttl converter ![breakout board connection image](https://wiki.pine64.org/images/thumb/6/6d/UART-pinecil-breakout-board-testing30.png/800px-UART-pinecil-breakout-board-testing30.png)

**Comment by sorgelig** [comment]
I didn't mean that full-featured board. What i want to do is to use simple type-c male to pcb breakout board where all type-c signals are output. Then i plan to connect DC5525 and uart pins to that board. As i see from schematics B3 and B2 pins are for UART. So i hope i just need to use those pins without messing with other pins. Not sure if pinecil swaps type-c sides depending on cable sense or they are fixed to specific sides.

**Comment by sorgelig** [comment]
do i need to use a special firmware to get UART log? I don't want to setup development environment to compile firmware myself.

**Comment by Ralim** [comment]
All firmware builds have the uart enabled.

Just reading off the pins is all you need; no flipping of the pins though they are always the same with cable orientation

**Comment by sorgelig** [comment]
ok, i've made USB-C pass-through adapter with B2 pin connected to RX of FT232. Connected by putty with 2000000pbs 8N1. 
I don't see any log in console...

**Comment by alteiis** [comment]
Same problem with v2.21:
 - official Pinecil v2 bought at pine64 store
 - flashed with the last blisp

 Ok if Bluetooth off (need to start in debug for first boot to disable Bluetooth)

**Comment by jonatan-ivanov** [comment]
Would it make sense to add a warning with the link to this issue to the release notes so people who check it before flashing the new version can decide if they want to go ahead or wait?

**Comment by sorgelig** [comment]
well.. FT232 can go up to 1500000bps it seems... 2mbps is crazy speed for UART. I don't have that sipeed rv-debugger board which supposed to be able to use 2mbps.
So, probably i'm out of UART testing...

**Comment by sorgelig** [comment]
according to datasheet, baudrate on FT232 is calculated as:
bps = 3000000/(n+x) where min n = 2, min x = 0. So max bitrate is 3000000/2 = 1500000bps. 
Anyway, i've tried - no activity on UART. May be i'm doing something wrong or 2mbps is too much for FT232.

**Comment by sorgelig** [comment]
Which crazy guy decided such insane UART speed? 2mbps to output text? 115200 is standard industrial UART speed..

**Comment by River-Mochi** [comment]
> There's information for the Pinecil Breakout board in the [pine64 wiki]

:)   I added that to wiki recently since there are often questions on how to set up UART & Break out board.  photo taken in my back yard.  not an expert; just followed  Ralim's directions and made  photo to help people. 
There is another method  too if you don't own a Pinecil Breakout board that I can upload if people need  it, then you buy   a  USB-C  M-F  passthrough pcb from ebay  or  aliexpress and solder a header pins to  GND, BC2 and BC3 (typically cost $3). you still need a siPeed  or NuSOM or some other UART type board as well.

**Comment by gamelaster** [comment]
@sorgelig this baudrate is used by Bouffalo as default baudrate. It is quite trend nowadays to use such high baud rates.
If you have really FT232 (withour RL or H on the end), it is likely that it might not be able to support that high baudrate.

**Comment by River-Mochi** [comment]
> might not be able to support that high baudrate.

I can confirm this  what Gamiee is saying,  some people in Pinecil chat can't use their low buad rate devices .  I can confirm that siPeed ($5 aliexpress)  and nuSOM brand  ($13, amazon)  both worked  when I tested them.

**Comment by sorgelig** [comment]
> It is quite trend nowadays to use such high baud rates.

i don't agree. Standard rate 115200 is used almost everywhere. Probably you deal only with same company or group of ppl who tend to use such unreasonably high speed.

**Comment by River-Mochi** [comment]
If you don't own a Pinecil breakout board and cant wait for  shipping. 
could use  B3, B2, and B1 (GND) on this  M-F   passthrough board. 
![image](https://user-images.githubusercontent.com/97197236/230961778-db34e15a-bd28-4bf7-9d8b-74ddb54dc727.png)


From Lupyuen's article on BL706  https://lupyuen.github.io/articles/bl706
![image](https://user-images.githubusercontent.com/97197236/230963156-ff4d7c7d-2ba4-4c79-af50-a2f23e6f1c65.png)

For people who don't have the nice Pinecil break out board,  this is possible to do UART with this to Pinecil  with these cheap PCB, one person had to repair the one they got from  Ebay bc trace was defective. and his VCC was bridged to BC3 so he has to fix that. check with meter on this cheap PCBs.

![image](https://user-images.githubusercontent.com/97197236/230963968-b08167ec-5229-4d43-8a64-abfafdd227e8.png)

![image](https://user-images.githubusercontent.com/97197236/230964502-bf00a2ef-d52e-44fd-b6d6-a85ab646a6dd.png)

**Comment by PSandro** [comment]
> > I don't think i want to open iron, i have bad experience in opening TS100.
> 
> anyone who wants **EASY  trick to open Pinecil**  without breaking plastic  clips. 
> **https://www.youtube.com/watch?v=aK01V5DrrVk**
> 
> opening is not needed to get  UART  debug messages,  just need   PCB boards like above  and   a free program like  PUTTY, then watch  screen while using pinecil.   https://www.putty.org/


Let's not to get off-topic :) I think this info is greatly appreciated in the wiki. I hope we can keep the comments here on a minimum so that everybody has an overview.

So when I receive the Breakout Bord I'll start testing to find out where the pinecil gets stuck during boot and will share the uart log here (only got an FT232RL, fingers crossed that it works; otherwise I'll try and build a firmware with lower baudrate).

@River-Mochi you seem to have the hardware needed to get the UART logs of your pinecilv2, can you share them for the blackscreen/chip crash scenario?

**Comment by chiefos** [comment]
@Ralim, it may be acceptable workaround to resolve the issue by adding in FW some delay to initialize BT only after Pinecil is already powered up?

**Comment by jonatan-ivanov** [comment]
I'm not really on top of all of these comments but would it make sense to:
1. Add a warning with the link to this issue to the release notes so people who check it before flashing the new version can decide if they want to go ahead or wait?
2. Releasing `2.21.1` or `2.22` (depending on versioning strategy) where BT is disabled by default (assuming that's the issue) and also making a comment in the new Release Notes that it might not work with the link of this issue?

**Comment by PSandro** [comment]
So I did some testing with building firmware, commenting some lines of code out and found the line of code where the blackscreens start happening:
https://github.com/Ralim/IronOS/blob/da18b9b60f6b4f4f9b23a03f6c5366e79d1a978b/source/Core/BSP/Pinecilv2/ble.c#L27

basically when I comment out this line and the following `hci_driver_init();` then there are no blackscreens/crashes on boot (of course there's also no bluetooth functionality).
I don't know anything about FreeRTOS but in the ble example of bouffalo_sdk, [they call the function](https://github.com/bouffalolab/bouffalo_sdk/blob/4b4e4c95449e902206e0d92025c6dc05f3e355d5/examples/ble/ble_peripheral/main.c#L211) using `xTaskCreateStatic` - does that maybe make a difference?

Also, bouffalo_sdk v2.0.0 was released two weeks ago and they changed the examples for ble a bit. One thing that's immediately noticeable: There's something called `rfparam_init`, maybe this needs to be done before initializing the ble_controller?

![image](https://user-images.githubusercontent.com/20563761/231293546-3fd6a169-3798-4ff9-b0cd-9d387b671693.png)

I don't know if bouffalo_sdk v2.0.0 can be integrated into this code or if there's even a plan to do so. Maybe this would resolve this issue? :smile:

**Comment by howels** [comment]
Same here after update to 2.21, black screen unless button held during the boot.

**Comment by River-Mochi** [comment]
more people in pinecil  live chat reporting black screen when updated to  2.21

**Comment by River-Mochi** [comment]
@Ralim 
UART testing: since the screen is either good when I plug it in (most of the time) or black. I'm not sure how to set up UART exactly for what you need collected.

in the past for UART,  I connected to windows 11,  opened device manager, find the com port pinecil uart device connected to and then open Putty to that specific comm port.  I have it connected now and I see data on putty  when I do various things with pinecil and pineSAM. I don't think this is what is wanted though?

Sample Putty screen showing UART  comm:
<img src="https://user-images.githubusercontent.com/97197236/231641917-8f245a25-1972-4171-b6dd-ed328a7f916f.png" width=600>

### I need more info on: 
1.  how exactly to set up for this specific test  to get data that would be helpful? (not the physcial plug into UART tester hardware, I have all that set up).
2.  is there a way to send this all the putty to a log file so I don't have to watch it and I just save a log somehow in  putty to submit to you? do I have to actively watch the putty screen?
3.  If I plug in pinecil and I don't have a black screen then is there any point to collecting UART info from Putty in that case?
4.  is there a special build of FW that could be loaded on pinecil  to get the better UART data that is needed?
5.  since it's either black when I plug in or it's all fine  then  how to best capture the correct info needed in uart?  so I'm not just looking watching UART info that is not needed?
6.  Do I just hook up the tester and keep unpluging the cable to the charger until the boot shows screen is black and then you want a screen capture of what is on Putty?

**Comment by PSandro** [comment]
I receive the breakout board today and could observe the UART while the pinecil failed to boot:
Here's the full output:
```
show flash cfg:
jedec id   0x000000
mid            0xC2
iomode         0x11
clk delay      0x01
clk invert     0x01
read reg cmd0  0x05
read reg cmd1  0x00
write reg cmd0 0x01
write reg cmd1 0x00
qe write len   0x02
cread support  0x00
cread code     0x00
burst wrap cmd 0x77
-------------------
Enable IRQ's
Pine64 Pinecilv2 Starting
BLE Starting
            Trap_Handler
mcause=30000007
mepc:23022800
mtval:00000000
Store/AMO access fault
```

**Comment by PlazmaZero** [comment]
I also got the black screen updating to 2.21 and disabling Bluetooth also fixed it for me.

**Comment by Ralim** [comment]
@PSandro Thank you for this,
That is indeed actually really useful; I suspect its faulting inside the BLE stack somewhere.
I'll try first making it be a delayed start to see if that helps.
If not will have to figure out more in depth answer.

**Comment by PSandro** [comment]
@Ralim could you reproduce this on your Pinecil v2 yet? I still don't understand why it only happens to some people and only on some boots... For me it faults in roughly 70% of the boots.
I looked up the instruction of the pc where it faults. If I'm not mistaken then it's inside some malloc, which itself is probably called within the BLE stack. Maybe there's not enough heap?

For the delaying: I tried delaying the BLE initialization (up to 5 seconds) and even started it in a dedicated thread. The fault still happens.

**Comment by sj-louw** [comment]
I have compiled the latest (as of 2023-04-23) branch with increased HEAP here https://github.com/Ralim/IronOS/blob/5faa092eb6b3c570ae48557030092883748a8a20/source/Core/BSP/Pinecilv2/FreeRTOSConfig.h#L17 to `((size_t)1024 * 14)`. Also I have applied this fix https://github.com/bouffalolab/bouffalo_sdk/issues/30 in `hal_mtimer.c` and cannot reproduce the black screen anymore. To me it seems to work OK now. Maybe someone else can also test my pre-compiled `Pinecilv2_EN.bin`, attached?

[Pinecilv2_EN.bin.zip](https://github.com/Ralim/IronOS/files/11304357/Pinecilv2_EN.bin.zip)

**Comment by PSandro** [comment]
> I have compiled the latest (as of 2023-04-23) `dev` branch with increased total HEAP
> 
> https://github.com/Ralim/IronOS/blob/5faa092eb6b3c570ae48557030092883748a8a20/source/Core/BSP/Pinecilv2/FreeRTOSConfig.h#L17
> 
> from `((size_t)1024 * 8)` to `((size_t)1024 * 14)`. Also, I have applied this fix [bouffalolab/bouffalo_sdk#30](https://github.com/bouffalolab/bouffalo_sdk/issues/30) in [`hal_mtimer.c`](https://github.com/Ralim/IronOS/blob/1fbcdcdf987a28541524e4f23ddc60fc942633ea/source/Core/BSP/Pinecilv2/bl_mcu_sdk/drivers/bl702_driver/hal_drv/src/hal_mtimer.c#L96-L100) and cannot reproduce the black screen anymore. To me, it seems to work OK now. Maybe someone else can also test the pre-compiled `Pinecilv2_EN.bin`, attached?
> [Pinecilv2_EN.bin.zip](https://github.com/Ralim/IronOS/files/11304357/Pinecilv2_EN.bin.zip)

Hi, I tried the binary you provided, sadly for me the blackscreen / `Store/AMO access fault` still happens. Also tried building a firmware with the patches applied that you mentioned - same issue.

Is there anything I can do to help debugging this? I wanted to get a backtrace through JTAG using OpenOCD and GDB. Unfortunately I don't know how to debug FreeRTOS applications with multiple Threads with that setup, my breakpoint on `bl_stack_start` never gets reached. Anyways... the `mepc` at fault points to an instruction inside `malloc_r`. I don't know what to do with that info - has this something to do with newlib and FreeRTOS?

**Comment by sj-louw** [comment]
@PSandro Not sure, but I have found this: https://nadler.com/embedded/newlibAndFreeRTOS.html

**Comment by sj-louw** [comment]
> Hia, I tried adjusting the linkerfile to try and have some more ram allocated to the heap. If you could give the `dev` builds a spin would be great. If that doesnt fix it, I'll try increasing the freertos heap as well.

Hi @Ralim I have tried your memory-tweaked build, but now the PinecilV2 is in a reboot loop. I have reverted again for now.

**Comment by NalleBerg** [comment]
> **Describe the bug** After flashing by blisp to v2.21 my pinecil v2 looks completely dead.
> 
> **To Reproduce**
> 
> 1. Flash v2.21 (Pinecilv2_EN.bin)
> 2. iron looks dead, no display on screen
> 
> **Expected behavior** normal work
> 
> **Details of your device:**
> 
> * Device: Pinecil v2 bought on pine64 site, so should be original one.
> * Release: i don't know how to see. in flash mode it doesn't show anything on screen.
> * Power adapter being used: tried both Type-C and 5525
> 
> **Additional context** v2.21 it seems doesn't work on my Pinecil v2. Thankfully, flashing v2.20 brings iron back to life.

I had the same problem. I solved it by flashing it twice. First I set it in programming mode by holding the [-]-button while inserting the USB C cable, attached to my PC and holding it for 15 sec.
Then the screen is black, but the PC recognises the Pinecil.
After that I flashed it with the command: .\blisp.exe write -c bl70x --reset .\Pinecilv2_NB.bin
I got a «Flash failed» message on the command line.
Then, without removing the USB-cable, I flashed one more time and then it flashed v2.21 just fine and I am now using it with the new firmware just fine.

It might have been just a fluke, but it worked for me; maybe it'll work for you too?

**Comment by aguilaair** [comment]
@Ralim mind providing the .bin artifact? Unfortunately, the GH Action artifacts are not visible to the public :(

**Comment by discip** [comment]
@aguilaair
Try this: [Refactor PinecilV2 Tuning](https://github.com/Ralim/IronOS/actions/runs/5166720294)

**Comment by rcmurphy** [comment]
This certainly fixed it on my brand new Pinecil v2 (I thought I was just being an idiot until I checked the issues 😅 )

**Comment by River-Mochi** [comment]
> Flashed 2.20, then that update directly on top of it, and now its happy? 🤷‍♀️
@rcmurphy 
Good news you got 2.21 working with the beta fix for 2.21

1.  did you need to update to 2.20 first and then  this 2.21  beta second to get it to work?
2.  did you use Blisp or Pineflash to do the update?
3. does this beta 2.21  work for you with Bluetooh  "ON"?
4. for example  bluetooth on and you are also able to use one of the pinecil  [bluetooth apps here](https://wiki.pine64.org/wiki/Pinecil_Firmware#Bluetooth_(BLE)_Apps) (e.g., joric's webpage graph or  builder555's  PineSAM) ?

**Comment by rcmurphy** [comment]
Yes, it all works, but it seems pretty unstable- I've unplugged it and had to reflash to get it working again.

**Comment by aguilaair** [comment]
> @aguilaair Try this: [Refactor PinecilV2 Tuning](https://github.com/Ralim/IronOS/actions/runs/5166720294)

After testing it, trying 2.20, and flashing this, I still experience the same symptoms. I can get it to eventually boot by unplugging it and plugging it back in, in which case it works, and Bluetooth is working.

**Comment by Ralim** [comment]
Any chance anyone who has crashing on newer builds also has a way to get the crash information from the UART?

**Comment by vadimcreates** [comment]
Hi, have the same issue with Pinecil V2 on fw 2.21. I had to turn off Bluetooth to make it working normally otherwise screen is black. @Ralim would provide more diag. info if you instruct me how to do it :)

**Comment by Ralim** [comment]
Diagnostics info comes out the uart connection, if you have a breakout that lets you view the uart that makes it easy.
(uart is pinned out on the extra USB3 pins). 

If you dont have the hardware handy for that dont panic. Other ways to help include trialling my attempts to figure out the issue.

On that, I've made this draft PR for increasing the stack size of the thread that starts BLE; if anyone could give it a spin and report back would be great. https://github.com/Ralim/IronOS/pull/1706

**Comment by aguilaair** [comment]
#1706 has the same issues. Unfortunately, I do not have a breakout right now

**Comment by vadimcreates** [comment]
> Diagnostics info comes out the uart connection, if you have a breakout that lets you view the uart that makes it easy. (uart is pinned out on the extra USB3 pins).
> 
> If you dont have the hardware handy for that dont panic. Other ways to help include trialling my attempts to figure out the issue.
> 
> On that, I've made this draft PR for increasing the stack size of the thread that starts BLE; if anyone could give it a spin and report back would be great. #1706

@Ralim I think I have the hardware but I don't know how to connect the pins. Should TX2+(B2) and TX2-(B3) from USB-C be connected with RX from USB to TTL adapter? Or the setup is different? I have putty installed that can easily connect to my USB-TTL adapter. If you can just give me a clue how to set it up, I will try to extract the logs for you.
![20230614_193204](https://github.com/Ralim/IronOS/assets/136507226/f5ad3dc9-267e-4f2c-aa86-c1126d561f8b)

**Comment by PSandro** [comment]
Hey, here's the output of my pinecil v2 (still) having the issue. I've flashed the Pinecilv2_EN.bin from the artifacts of dev build. SHA256 sum of the bin file is `296bd431a5273ae9752de10261373e6c444da29d76760e7fc0c68384693cf9e1`.

I confirmed on the pinecil, that it is indeed version `v2.21.CAA638C` built on 14-06-23. Please let me know if I can be of any further help in fixing this :)

```
show flash cfg:
jedec id   0x000000
mid            0xC2
iomode         0x11
clk delay      0x01
clk invert     0x01
read reg cmd0  0x05
read reg cmd1  0x00
write reg cmd0 0x01
write reg cmd1 0x00
qe write len   0x02
cread support  0x00
cread code     0x00
burst wrap cmd 0x77
-------------------
Enable IRQ's
Pine64 Pinecilv2 Starting
BLE Starting
            Trap_Handler
mcause=30000007
mepc:23023dd6
mtval:00000000
Store/AMO access fault
```

**Comment by Ralim** [comment]
alrighty I pushed up a different solution that would be great for someone to test if they could

**Comment by PSandro** [comment]
Thanks for the updated, I can still reproduce the bug. But now it seems to happen less often.
sha256sum: `25d9058c6625103706aaf090d0ca2cc87c18521b1778ed5e4562a383b8e5553f  Pinecilv2_EN.bin`
version: `v2.21.4EF3584`

output:
```
show flash cfg:
jedec id   0x000000
mid            0xC2
iomode         0x11
clk delay      0x01
clk invert     0x01
read reg cmd0  0x05
read reg cmd1  0x00
write reg cmd0 0x01
write reg cmd1 0x00
qe write len   0x02
cread support  0x00
cread code     0x00
burst wrap cmd 0x77
-------------------
Enable IRQ's
Pine64 Pinecilv2 Starting
BLE Starting
            Trap_Handler
mcause=30000007
mepc:23024074
mtval:00000000
Store/AMO access fault
```

**Comment by sorgelig** [comment]
I wonder why Pine64 cannot help with nailing the issue. They have many pinecils and definitely can find those not working after update and then test it thoughtfully or even send it to Ralin. Didn't Ralim advertised them as most responsive and community oriented team? 3 months have passed and source of problem is still unclear.

**Comment by Ralim** [comment]
Thank you @PSandro 

That is excellent to hear, I suspect its still and out-of ram condition.
I might upload some extra firmware for you to test.
To answer a question I've been wondering, any chance you could scan and see how many ble devices are nearby to you?
I'm wondering if this is either _it_ doing a scan and OOM'ing or some random device querying something it cant handle.


---

@sorgelig 

So there are a few things that need to be set straight before we go into this and before anyone starts yelling at Pine.
(1) 
BLE is _not_ a Pine advertised feature and in all design work it is not a core function of the device. This is the reason its not on the store page (Which is run by Pine) but mostly only on community pages. This decision as far as I understand was because the BL70x is a young chip on an unstable platform, and trust is not yet built up with buffalo that its bug free (for reference it definitely has bugs).

The people who should be in question for helping here are buffalo but I'll get to them in a second.

(2)
Thus all BLE is down to me deciding to build it because of community asking for it. It took a while to take the very poor `mcu_sdk` release at the time and cull it down enough that (1) I could ship it and (2) it sort of worked. It was always going to be a second tier goal / priority as its not required for the operation of the device at its core. and coupled with (1) It not being an advertised fully supported feature.

(3)
One can note that Buffalo no longer supports BLE or even the radio at all in the bl70x in all of the newer revisions of the SDK.
I'm running on an older SDK just so we can have BLE, even though it means its a mess for support; and will eventually become its own nightmare.

(4)
Pine64 have been _great_ for communication and support, and its not like they don't know about this. 
But they cant really do anything here. Its down to Buffalo at this point. 
There are lots of other aspects of Pinecil that they do help with _often_ but this is not one they have any control over. Buffalo have decided on their current trajectory and as of present it doesn't look to include much for the BL70x. The vibe is that they are focused on other devices at this point in time.

(5)
Additionally its been quite hard to get reliable reproductions of it, as some tests factory side didn't show the issue, nor have any of the devices I have here or other Pine staff. And yes these are across multiple batches of devices. And I've already confirmed with pine its same BL702 part number being fitted etc.


----

So at the moment given the above, its a bit of a goose chase to debug what is OOM'ing and not handling itself correctly.

I have narrowed it down to likely areas. If anyone in this thread is comfortable editing the code and re-flashing a bunch to try and narrow down the code I'd love to pair up with you to gain more insight.. Otherwise I'll continue my slow binary search of potential causes. I've found ways to definitely cause this to occur but they still dont 1:1 match up so its slow.


Also keep in mind that yes "3 months have passed and source of problem is still unclear", but also in those 3 months I have not been around much to actually do anything. As this is a project of "when time is available", there are long stretches where I do not open it _all_. Ive only really spend a few hours on this fragmented over all of that time.

**Comment by sorgelig** [comment]
Since BLE is an unavoidable part of newer firmware i suggest to turn BLE off by default. More over, it would be wise to clear this flag with EVERY flash of firmware. In that case, if some pinecil has this problem then simply reflash will fix. Those, who have no such issue and would like to use BLE will simply re-enable it after flash. So, firmware can move forward fixing other bugs while this issue can live some time without distracting.

**Comment by gamelaster** [comment]
Since it is hard to remotely figure out why the crash happens, I can offer someone an fully working Pinecil V2, in exchange of Pinecil V2, which crashes on v2.21 . Shipment would go from/to Slovakia, I will cover the shipment fees. If you are interested, let me know by email on gamiee (at) pine64.org . Thanks!

**Comment by aguilaair** [comment]
I'd be happy to ship mine, but it has the transparent case, any chance you have one that also has it?

**Comment by aguilaair** [comment]
I'm currently in Spain!

**Comment by gamelaster** [comment]
@aguilaair ohhh, great! Can you please contact me on Discord (gamelaster), Telegram: ( gamiee ) or on email gamiee (at) pine64.org ? Thanks!

**Comment by aguilaair** [comment]
Sent, @gamelaster!

**Comment by gamelaster** [comment]
Probably closed by accident.  I received affected Pinecil V2 from @aguilaair (once again big thanks!), and I will try to look on it ASAP. Hopefully we can fix it soon.

**Comment by rcmurphy** [comment]
Let me know if you need another!I’m in the US.Best,RebeccaOn Jul 20, 2023, at 14:34, Marek Kraus ***@***.***> wrote:﻿
Probably closed by accident.  I received affected Pinecil V2 from @aguilaair (once again big thanks!), and I will try to look on it ASAP. Hopefully we can fix it soon.

**Comment by Ralim** [comment]
Hello 🙇🏼 

Would like to update here that I've posted a PR which has some new build assets to test at #1756.
Would be amazing if we could get some more testing by anyone affected or not affected.
The hope is that this (1) resolves the crash, and (2) gets BLE working.
I'll take (1) but realllllllly hoping for (2).

**Comment by gamelaster** [comment]
@sorgelig pleaes can you try? Thanks!

**Comment by sorgelig** [comment]
Sure, but where to get a ready to flash firmware?

**Comment by gamelaster** [comment]
@sorgelig https://github.com/Ralim/IronOS/actions/runs/5635830175?pr=1756 , in PinecilV2.zip you should have pre-built binary for test.

**Comment by sorgelig** [comment]
flashed already.  It seems booting fine with BT enabled!
Now i need to find that android app for testing. 
Attempt for pairing from android BT settings says "An app is needed to use this device".

**Comment by gamelaster** [comment]
@sorgelig yes, app is needed, there is one made by @aguilaair but not published yet.

**Comment by sorgelig** [comment]
Just tested on Windows with PineSAM from post above. It works fine!
I can close this issue, or keep it opened till official release?

**Comment by sorgelig** [comment]
Great work @Ralim!

**Comment by vadimcreates** [comment]
Well done! Just flashed mine and tested. It works normal.

**Comment by gamelaster** [comment]
@vadimcreates yay! Amazing to hear.
@sorgelig Great to hear that PineSAM also works :partying_face: . And no need to close it, we will do it when patch gets merged into PR.

Thank you all for tests! :+1:

**Comment by hasheddan** [comment]
I found this thread very informative when I was looking at some of the internals of the Pinecil v2 support in IronOS. Since there was some discussion of getting access to the UART for logs on the Pinecil v2, which causes this thread to surface for various related search terms, I thought I would drop a link to the setup I've been using with a Raspberry Pi Pico. Hope its helpful for folks who might have a few Pico's lying around! https://danielmangum.com/posts/risc-v-bytes-accessing-pinecil-uart-picoprobe/

**Comment by gamelaster** [comment]
That's amazing! Thanks :)
